In [3]:
from llama_index.core import SimpleDirectoryReader

from llama_index.readers.file import PDFReader, PyMuPDFReader

In [10]:
# PDF Reader with 'SimpleDirectoryReader'
parser = PDFReader()
file_extractor = {'.pdf': parser}
documents = SimpleDirectoryReader("../../Data/input", file_extractor = file_extractor).load_data()
documents[0].text

"7/1/23, 1 1:31 PM The World's Billionaires - Wikipedia\nhttps://en.wikipedia.or g/wiki/The_W orld%27s_Billionaires 1/33The World's Billionaires\nList of the world's billionaires, ranked in order of net worth\nThe net worth of the world's billionaires increased from\nless than US$1 trillion in 2000 to over $7 trillion in 2015.\nPublication details\nPublisher Whale Media Investments\nForbes family\nPublication F o r b e s\nFirst published March 1987[ 1 ]\nLatest publication April 4, 2023\nCurrent list details (2023)[ 2 ]\nWealthiest Bernard Arnault\nNet worth (1st)\n \xa0US$211\xa0billion\nNumber of\nbillionaires\n\xa02,640 (from 2668)\nTotal list net worth\nvalue\n\xa0US$12.2\xa0trillion (from US$ 12.7\ntrillion)\nNumber of women\n \xa0337\nNew members to the\nlist\n\xa0150\nF o r b e s: The World's Billionaires website (https://www.forb\nes.com/billionaires/)The World's Billionaires\nThe World's Billionaires  is an annual\nranking of people who are considered to have a\nnet worth  of 

In [8]:
# PyMuPDF Reader example
parser = PyMuPDFReader()
file_extractor = {".pdf": parser}
documents = SimpleDirectoryReader("../../Data/input", file_extractor = file_extractor).load_data()
documents[0].text

"7/1/23, 11:31 PM\nThe World's Billionaires - Wikipedia\nhttps://en.wikipedia.org/wiki/The_World%27s_Billionaires\n1/33\nThe World's Billionaires\nList of the world's billionaires, ranked in order of net worth\nThe net worth of the world's billionaires increased from\nless than US$1 trillion in 2000 to over $7 trillion in 2015.\nPublication details\nPublisher\nWhale Media Investments\nForbes family\nPublication\nForbes\nFirst published\nMarch 1987[1]\nLatest publication\nApril 4, 2023\nCurrent list details (2023)[2]\nWealthiest\nBernard Arnault\nNet worth (1st)\n\xa0US$211\xa0billion\nNumber of\nbillionaires\n\xa02,640 (from 2668)\nTotal list net worth\nvalue\n\xa0US$12.2\xa0trillion (from US$ 12.7\ntrillion)\nNumber of women\n\xa0337\nNew members to the\nlist\n\xa0150\nForbes: The World's Billionaires website (https://www.forb\nes.com/billionaires/)\nThe World's Billionaires\nThe World's Billionaires is an annual\nranking of people who are considered to have a\nnet worth of $1 billion

### Samrt PDF Loader 

SmartPDFLoader is a super fast PDF reader that understands the layout structure of PDFs such as nested sections, nested lists, paragraphs and tables. It uses layout information to smartly chunk PDFs into optimal short contexts for LLMs.

In [1]:
from llama_index.readers.smart_pdf_loader import SmartPDFLoader
#llama-index-readers-smart-pdf-loader
#llmsherpa

In [2]:
llmsherpa_api_url = "https://readers.llmsherpa.com/api/document/developer/parseDocument?renderFormat=all"
pdf_url = "../../Data/input/billionaires_page.pdf"

pdf_loader = SmartPDFLoader(llmsherpa_api_url = llmsherpa_api_url)
documents = pdf_loader.load_data(pdf_url)

In [5]:
documents[0].text

"\n | 7/1/23, 11:31 PM | The World's Billionaires - Wikipedia\n | --- | ---\n | The World's Billionaires\n | The World's Billionaires is an annual ranking of people who are considered to have a net worth of $1 billion or more, by the American business magazine Forbes. The list was first published in March 1987.[1] The total net worth of each individual on the list is estimated and is cited in United States dollars, based on their documented assets and accounting for debt and other factors. Royalty and dictators whose wealth comes from their | The World's Billionaires\n | List of the world's billionaires, ranked in order of net worth\n | positions are excluded from these lists.[3] This ranking is an index of the wealthiest documented individuals, excluding any ranking of those with wealth that is not able to be completely ascertained.[4] | The net worth of the world's billionaires increased from less than US$1 trillion in 2000 to over $7 trillion in 2015.\n | Publication details In 2018

### Llama Parse

In [2]:
import nest_asyncio

nest_asyncio.apply()

In [ ]:
from llama_parse import LlamaParse

parser = LlamaParse(
    api_key = 'llx-MTwEiTV3Y4hLciDAX5n5nEi84oQcAJSUxtbAY2Dr3OZDih5p', # can also be set in your env as LLAMA_CLOUD_API_KEY
    result_type = "markdown", # "markdown" and "text" are available
    verbose = True
)

#sync 
documents = parser.load_data("../../Data/input/billionaires_page.pdf") #["./my_file1.pdf", "./my_file2.pdf"]
documents

In [6]:
documents[0].text

'# The World\'s Billionaires\n\nThe World\'s Billionaires is an annual ranking of people who are considered to have a net worth of $1 billion or more, by the American business magazine Forbes. The list was first published in March 1987. The total net worth of each individual on the list is estimated and is cited in United States dollars, based on their documented assets and accounting for debt and other factors. Royalty and dictators whose wealth comes from their positions are excluded from these lists. This ranking is an index of the wealthiest documented individuals, excluding any ranking of those with wealth that is not able to be completely ascertained.\n\nThe net worth of the world\'s billionaires increased from less than US$1 trillion in 2000 to over $7 trillion in 2015.\n\n# Publication details\n\n|Publisher|Whale Media Investments|\n|---|---|\n|Publication|Forbes|\n|First published|March 1987|\n|Latest publication|April 4, 2023|\n\n# Current list details (2023)\n\n|Wealthiest|B

'# The World\'s Billionaires\n\nThe World\'s Billionaires is an annual ranking of people who are considered to have a net worth of $1 billion or more, by the American business magazine Forbes. The list was first published in March 1987. The total net worth of each individual on the list is estimated and is cited in United States dollars, based on their documented assets and accounting for debt and other factors. Royalty and dictators whose wealth comes from their positions are excluded from these lists. This ranking is an index of the wealthiest documented individuals, excluding any ranking of those with wealth that is not able to be completely ascertained.\n\nThe net worth of the world\'s billionaires increased from less than US$1 trillion in 2000 to over $7 trillion in 2015.\n\n# Publication details\n\n|Publisher|Whale Media Investments|\n|---|---|\n|Publication|Forbes|\n|First published|March 1987|\n|Latest publication|April 4, 2023|\n\n# Current list details (2023)\n\n|Wealthiest|Bernard Arnault|\n|---|---|\n|Net worth (1st)|US$211 billion|\n|Number of billionaires|2,640 (from 2668)|\n|Total list net worth|US$12.2 trillion (from US$ 12.7 trillion)|\n|Number of women|337|\n|New members to the list|150|\n\nIn 2018, Amazon founder Jeff Bezos was ranked at the top for the first time and became the first centibillionaire included in the ranking, surpassing Microsoft founder Bill Gates, who had topped the list 18 of the previous 24 years. In 2022, after topping the list for four years, Bezos was surpassed by Elon Musk. In 2023, Musk was in turn surpassed by French businessman Bernard Arnault, after topping the list for just a year. Arnault became the first French person to top the list.\n\n# Methodology\n\nEach year, Forbes employs a team of over 50 reporters from a variety of countries to track the activity of the world\'s wealthiest individuals and sometimes groups or families who share wealth. Preliminary surveys are sent to those who may qualify for the list. According to Forbes, they received three types of responses – some people try to inflate their wealth, others cooperate but leave out details, and some refuse to answer any questions. Business deals are then scrutinized and estimates of valuable assets – land, homes, vehicles, artwork, etc. – are made. Interviews are conducted to vet the figures and improve the estimate of an individual\'s holdings. Finally, positions in a publicly traded.\n---\nThe World\'s Billionaires - Wikipedia\n\nstock are priced to market on a date roughly a month before publication. Privately held companies are priced by the prevailing price-to-sales or price-to-earnings ratios. Known debt is subtracted from assets to get a final estimate of an individual\'s estimated worth in United States dollars. Since stock prices fluctuate rapidly, an individual\'s true wealth and ranking at the time of publication may vary from their situation when the list was compiled.\n\nWhen a living individual has dispersed his or her wealth to immediate family members it is included under a single listing (as a single "family fortune") provided that individual (the grantor) is still living. However, if a deceased billionaire\'s fortune has been dispersed, it will not appear as a single listing, and each recipient will only appear if his or her own total net worth is over a $Billion (his or her net worth will not be combined with family members\'). Royal families and dictators that have their wealth contingent on a position are always excluded from these lists.\n\nAnnual rankings\n\nThe rankings are published annually in March, so the net worths listed are snapshots taken at that time. These lists only show the top 10 wealthiest billionaires for each year.\n\nLegend\n\n|Icon|Description|\n|---|---|\n|Has not changed from the previous ranking.| |\n|Has increased from the previous ranking.| |\n|Has decreased from the previous ranking.| |\n\n2023\n\nIn the 37th annual Forbes list of the world\'s billionaires, the list included 2,640 billionaires with a total net wealth of $12.2 trillion, down 28 members and $500 billion from 2022. Nearly half the list is poorer than the previous year, including Elon Musk, who fell from No. 1 to No. 2. The list also marks for the first time a French citizen was in the top position as well as a non-American for the first time since 2013 when the Mexican Carlos Slim Helu was the world\'s richest person. The list, like in 2022, counted 15 under 30 billionaires with the richest of them being Red Bull heir Mark Mateschitz with a net worth of $34.7 billion. The youngest of the lot were Clemente Del Vecchio, heir to the Luxottica fortune shared with his six siblings and stepmother and Kim Jung-yang, whose fortune lies in Japanese-South Korean gaming giant Nexon, both under-20s.\n---\n# The World\'s Billionaires - Wikipedia\n\n|No.|Name|Net worth (USD)|Age|Nationality|Primary source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bernard Arnault & family|$211 billion|74|France|LVMH|\n|2|Elon Musk|$180 billion|51|United States|Tesla, SpaceX, X Corp.|\n|3|Jeff Bezos|$114 billion|59|United States|Amazon|\n|4|Larry Ellison|$107 billion|78|United States|Oracle Corporation|\n|5|Warren Buffett|$106 billion|92|United States|Berkshire Hathaway|\n|6|Bill Gates|$104 billion|67|United States|Microsoft|\n|7|Michael Bloomberg|$94.5 billion|81|United States|Bloomberg L.P.|\n|8|Carlos Slim & family|$93 billion|83|Mexico|Telmex, América Móvil, Grupo Carso|\n|9|Mukesh Ambani|$83.4 billion|65|India|Reliance Industries|\n|10|Steve Ballmer|$80.7 billion|67|United States|Microsoft|\n\nIn the 36th annual Forbes list of the world\'s billionaires, the list included 2,668 billionaires with a total net wealth of $12.7 trillion, down 97 members from 2021.\n\nSource: Wikipedia\n---\n# The World\'s Billionaires - Wikipedia\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Elon Musk|$219 billion|50|United States|Tesla, SpaceX, Twitter Inc|\n|2|Jeff Bezos|$177 billion|58|United States|Amazon|\n|3|Bernard Arnault & family|$158 billion|73|France|LVMH|\n|4|Bill Gates|$129 billion|66|United States|Microsoft|\n|5|Warren Buffett|$118 billion|91|United States|Berkshire Hathaway|\n|6|Larry Page|$111 billion|49|United States|Alphabet Inc.|\n|7|Sergey Brin|$107 billion|48|United States|Alphabet Inc.|\n|8|Larry Ellison|$106 billion|77|United States|Oracle Corporation|\n|9|Steve Ballmer|$91.4 billion|66|United States|Microsoft|\n|10|Mukesh Ambani|$90.7 billion|64|India|Reliance Industries|\n\nIn the 35th annual Forbes list of the world\'s billionaires, the list included 2,755 billionaires with a total net wealth of $13.1 trillion, up 660 members from 2020; 86% of these billionaires had more wealth than they possessed last year.\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Jeff Bezos|$177 billion|57|United States|Amazon|\n|2|Elon Musk|$151 billion|49|United States|Tesla, SpaceX|\n|3|Bernard Arnault & family|$150 billion|72|France|LVMH|\n|4|Bill Gates|$124 billion|65|United States|Microsoft|\n|5|Mark Zuckerberg|$97 billion|36|United States|Meta Platforms|\n|6|Warren Buffett|$96 billion|90|United States|Berkshire Hathaway|\n|7|Larry Ellison|$93 billion|76|United States|Oracle Corporation|\n|8|Larry Page|$91.5 billion|48|United States|Alphabet Inc.|\n|9|Sergey Brin|$89 billion|47|United States|Alphabet Inc.|\n|10|Mukesh Ambani|$84.5 billion|63|India|Reliance Industries|\n---\n# The World\'s Billionaires - Wikipedia\n\nIn the 34th annual Forbes list of the world\'s billionaires, the list included 2,095 billionaires with a total net wealth of $8 trillion, down 58 members and $700 billion from 2019; 51% of these billionaires had less wealth than they possessed last year. The list was finalized as of 18 March, thus was already partially influenced by the COVID-19 pandemic.\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Jeff Bezos|$113 billion|56|United States|Amazon|\n|2|Bill Gates|$98 billion|64|United States|Microsoft|\n|3|Bernard Arnault & family|$76 billion|71|France|LVMH|\n|4|Warren Buffett|$67.5 billion|89|United States|Berkshire Hathaway|\n|5|Larry Ellison|$59 billion|75|United States|Oracle Corporation|\n|6|Amancio Ortega|$55.1 billion|84|Spain|Inditex, Zara|\n|7|Mark Zuckerberg|$54.7 billion|35|United States|Facebook, Inc.|\n|8|Jim Walton|$54.6 billion|71|United States|Walmart|\n|9|Alice Walton|$54.4 billion|70|United States|Walmart|\n|10|S. Robson Walton|$54.1 billion|77|United States|Walmart|\n\nIn the 33rd annual Forbes list of the world\'s billionaires, the list included 2,153 billionaires with a total net wealth of $8.7 trillion, down 55 members and $400 billion from 2018. The U.S. continued to have the most billionaires in the world, with a record of 609, while China dropped to 324 (when not including Hong Kong, Macau and Taiwan).\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Jeff Bezos|$131 billion|55|United States|Amazon|\n|2|Bill Gates|$96.5 billion|63|United States|Microsoft|\n|3|Warren Buffett|$82.5 billion|88|United States|Berkshire Hathaway|\n|4|Bernard Arnault|$76 billion|70|France|LVMH|\n|5|Carlos Slim|$64 billion|79|Mexico|América Móvil, Grupo Carso|\n|6|Amancio Ortega|$62.7 billion|82|Spain|Inditex, Zara|\n|7|Larry Ellison|$62.5 billion|74|United States|Oracle Corporation|\n|8|Mark Zuckerberg|$62.3 billion|34|United States|Facebook, Inc.|\n|9|Michael Bloomberg|$55.5 billion|77|United States|Bloomberg L.P.|\n|10|Larry Page|$50.8 billion|45|United States|Alphabet Inc.|\n---\n# The World\'s Billionaires - Wikipedia\n\nIn the 32nd annual Forbes list of the world\'s billionaires, the aggregate wealth of the top 20 richest people on Earth amounted to about 13 percent of all billionaires\' fortunes combined. A record of 2,208 billionaires were in the ranking and the total wealth was $9.1 trillion, up 18% since 2017. For the first time, Jeff Bezos was listed as the top billionaire due to Amazon\'s rising stock price that resulted in one person\'s biggest one-year gain in wealth ($35 billion) since Forbes started tracking in 1987. The U.S. had the most billionaires in the world, with 585, while China was catching up with 476 when including Hong Kong, Macau and Taiwan; it had 372 when excluding those three places. Forbes excluded Al-Walid bin Talal and all other Saudi billionaires due to the absence of accurate wealth estimations as a result of the 2017–19 Saudi Arabian purge.\n\n|No.|Name|Net worth (USD)|\n|---|---|---|\n|1|Jeff Bezos|$112 billion|\n|2|Bill Gates|$90 billion|\n|3|Warren Buffett|$84 billion|\n|4|Bernard Arnault|$72 billion|\n|5|Mark Zuckerberg|$71 billion|\n\n|Age|Nationality|Source(s) of wealth|\n|---|---|---|\n|54|United States|Amazon|\n|62|United States|Microsoft|\n|87|United States|Berkshire Hathaway|\n|69|France|LVMH|\n|33|United States|Facebook, Inc.|\n|81|Spain|Inditex, Zara|\n|78|Mexico|América Móvil, Grupo Carso|\n|82|United States|Koch Industries|\n|77|United States|Koch Industries|\n|73|United States|Oracle Corporation|\n\nOn the 30th anniversary of the Forbes list of the world\'s billionaires, for the fourth year in a row, Bill Gates was named the richest man in the world. In 2017, there was a record of 2,043 people on the list, which is the first time over 2,000 people were listed. This included 195 newcomers of whom 76 were from China and 25 from the U.S.; there were 56 people under 40 and it had a record of 227 women. The number of billionaires increased 13% to 2,043 from 1,810 in 2016; this was the biggest change in over 30 years of tracking billionaires globally. Added together, the total net worth for 2017\'s billionaires was US$7.67 trillion, up from US$7.1 trillion in 2015. This was the first time after 12 years that Carlos Slim was not within the top five. The U.S. had the most billionaires in the world, with a record of 565. China had 319 (not including Hong Kong, Taiwan or Macau), Germany had 114, and India had the fourth most with 101; India reached over 100 billionaires for its first time.\n\nSource: https://en.wikipedia.org/wiki/The_World%27s_Billionaires\n---\n# The World\'s Billionaires - Wikipedia\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$86.0 billion|61|United States|Microsoft|\n|2|Warren Buffett|$75.6 billion|86|United States|Berkshire Hathaway|\n|3|Jeff Bezos|$72.8 billion|53|United States|Amazon|\n|4|Amancio Ortega|$71.3 billion|80|Spain|Inditex, Zara|\n|5|Mark Zuckerberg|$56.0 billion|32|United States|Facebook, Inc.|\n|6|Carlos Slim|$54.5 billion|77|Mexico|América Móvil, Grupo Carso|\n|7|Larry Ellison|$52.2 billion|72|United States|Oracle Corporation|\n|8|Charles Koch|$48.3 billion|81|United States|Koch Industries|\n|8|David Koch|$48.3 billion|76|United States|Koch Industries|\n|10|Michael Bloomberg|$47.5 billion|75|United States|Bloomberg L.P.|\n\nFor the third year in a row, Bill Gates was named the richest man in the world by Forbes\' 2016 list of the world\'s billionaires. This is the 17th time that the founder of Microsoft had claimed the top spot. Amancio Ortega rose from last year\'s position of number four to second. Warren Buffett of Berkshire Hathaway came in third for the second consecutive time, while Mexican telecommunication mogul Carlos Slim slipped from last year\'s second position to fourth. Jeff Bezos of Amazon, Mark Zuckerberg of Facebook, and Michael Bloomberg of Bloomberg L.P., appeared for the first time on the Forbes top 10 billionaires list, coming at fifth, sixth, and eighth positions, respectively. Zuckerberg became the youngest top 10 billionaire this year at the age of 31. Larry Ellison, Charles Koch, and David Koch also slipped from their previous year\'s positions, with Ellison dropping to seventh from fifth, and the Kochs falling to ninth position from sixth.\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$75.0 billion|60|United States|Microsoft|\n|2|Amancio Ortega|$67.0 billion|79|Spain|Inditex|\n|3|Warren Buffett|$60.8 billion|85|United States|Berkshire Hathaway|\n|4|Carlos Slim|$50.0 billion|76|Mexico|América Móvil, Grupo Carso|\n|5|Jeff Bezos|$45.2 billion|52|United States|Amazon|\n|6|Mark Zuckerberg|$44.6 billion|31|United States|Facebook, Inc.|\n|7|Larry Ellison|$43.6 billion|71|United States|Oracle Corporation|\n|8|Michael Bloomberg|$40.0 billion|74|United States|Bloomberg L.P.|\n|9|Charles Koch|$39.6 billion|80|United States|Koch Industries|\n|9|David Koch|$39.6 billion|75|United States|Koch Industries|\n---\n# The World\'s Billionaires - Wikipedia\n\nIn the 29th annual Forbes list of global billionaires, a record 1,826 billionaires were named with an aggregated net worth of $7.1 trillion compared to $6.4 trillion in the previous year. 46 of the billionaires in this list were under the age of 40. A record number of 290 people joined the list for the first time, of whom 25 percent hailed from China, which produced a world-leading 71 newcomers. The United States came in second, with 57; followed by India, with 28; and Germany, with 23. The United States had the largest number of billionaires with 526. Russia went down to 88 from 111 in 2014. Russia was placed behind China, Germany and India by the number of billionaires. Self-made billionaires made up the largest number of people on the list with 1,191 positions (over 65 percent), while just 230 (under 13 percent) had wealth through inheritance. The number of billionaires who inherited a portion but were still working to increase their fortunes is 405.\n\nBill Gates was named the richest man in the world by Forbes\' annual list of the world\'s billionaires. This was the 16th time that the founder of Microsoft claimed the top spot. Carlos Slim came in second for the second consecutive time. Warren Buffett of Berkshire Hathaway placed third, while Amancio Ortega of Spain slipped down a position from the previous year to number four. Larry Ellison, the founder of Oracle, rounded off the top five. Christy Walton was the highest-ranking female at number eight. America\'s Evan Spiegel, co-founder of photo messaging app Snapchat, became the youngest billionaire this year at age 24. At age 99, David Rockefeller maintained his position as the oldest billionaire included in the list. Mark Zuckerberg, the founder of Facebook, rose to number 16 with $33.4 billion. Iceland had a billionaire, Thor Bjorgolfsson, in the list after a gap of five years. Guatemala had a billionaire, Mario Lopez Estrada, for the first time in its history.\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$79.2 billion|59|United States|Microsoft|\n|2|Carlos Slim|$77.1 billion|75|Mexico|América Móvil, Grupo Carso|\n|3|Warren Buffett|$72.7 billion|84|United States|Berkshire Hathaway|\n|4|Amancio Ortega|$64.5 billion|78|Spain|Inditex|\n|5|Larry Ellison|$54.3 billion|70|United States|Oracle Corporation|\n|6|Charles Koch|$42.9 billion|79|United States|Koch Industries|\n|6|David Koch|$42.9 billion|74|United States|Koch Industries|\n|8|Christy Walton|$41.7 billion|66|United States|Walmart|\n|9|Jim Walton|$40.6 billion|66|United States|Walmart|\n|10|Liliane Bettencourt|$40.1 billion|92|France|L\'Oreal|\n\nGates added $9 billion to his fortune since 2013 and topped the Forbes 2014 billionaire list. He had topped the list in 15 of the previous 20 years, but his previous number one ranking was in 2009. Mexican telecommunication mogul Carlos Slim came in second place after being number one the previous four years. Zara founder Amancio Ortega placed third for the second consecutive year. American investor Warren Buffett was in the top five for the 20th consecutive year, placing fourth. America\'s Christy Walton was the highest ranking woman, placing ninth overall. Aliko Dangote of Nigeria became the first African to enter the top 25, with an estimated net worth of $25 billion.\n---\n# The World\'s Billionaires - Wikipedia\n\nA total of 1,645 people made the 2014 billionaire list, representing a combined wealth of $6.4 trillion.[30] Of those, a record 268 were newcomers, surpassing 2008\'s 226 newcomers.[1] 100 people listed in 2013 failed to make the list.[30] The number of women on the list rose to a record 172 in 2014.[1] Approximately 66 percent of the list were self-made, 13 percent achieved their wealth through inheritance alone, and 21 percent through a mixture of the two.[3]\n\nThe United States had 492 billionaires on the list, the most of any country.[30] It also had the most newcomers with 50, and women with 54.[1] China had the second most billionaires with 152, while Russia was third with 111.[30] Algeria, Lithuania, Tanzania, and Uganda were all represented on the list for the first time. Turkey saw the most people drop off the list, 19, due to a period of high inflation in the country.[3]\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$76.0 billion|58|United States|Microsoft|\n|2|Carlos Slim & family|$72.0 billion|74|Mexico|América Móvil, Grupo Carso|\n|3|Amancio Ortega|$64.0 billion|77|Spain|Inditex|\n|4|Warren Buffett|$58.2 billion|83|United States|Berkshire Hathaway|\n|5|Larry Ellison|$48.0 billion|70|United States|Oracle Corporation|\n|6|Charles Koch|$40.0 billion|78|United States|Koch Industries|\n|6|David Koch|$40.0 billion|73|United States|Koch Industries|\n|8|Sheldon Adelson|$38.0 billion|80|United States|Las Vegas Sands|\n|9|Christy Walton & family|$36.7 billion|65|United States|Walmart|\n|10|Jim Walton|$34.7 billion|65|United States|Walmart|\n\n2013\n\nCarlos Slim topped the 2013 billionaire list, marking his fourth consecutive year at the top. Gates remained in second, while Amancio Ortega moved up to third. Ortega\'s gain of $19.5 billion was the largest of anyone on the list.[31] Warren Buffett failed to make the top three for the first time since 2000, placing fourth. Diesel founder Renzo Rosso was among the top newcomers, debuting with an estimate net worth of $3 billion.[32]\n\nA global rise in asset prices led Forbes editor Randall Lane to declare "It [was] a very good year to be a billionaire". However, it was not a good year to be Eike Batista, who fell from seventh to 100th, suffering the largest net loss of anyone on the list.[31] Overall, net gainers outnumbered net losers by 4:1.[32]\n\nA record total of 1,426 people made the 2013 list, representing $5.4 trillion of assets.[33] Of those, 442 billionaires hailed from the United States.[31] The Asian-Pacific region had 386 billionaires and Europe 366. The list also featured a record number of newcomers, 210, representing 42 countries. 60 people from the 2012 list fell below a billion dollars of assets in 2013, and eight others from the 2012 list died. The Asia-Pacific region had the most drop-offs, with 29, followed by the United States with 16. The 2013 list featured 138 women, of which 50 came from the United States. A majority of the list\n---\n# The World\'s Billionaires - Wikipedia\n\n(961 individuals, 67 percent) were entirely self-made; 184 (13 percent) inherited their wealth, and 281\n(20 percent) achieved their fortune through a combination of inheritance and business acumen.[33]\nVietnam\'s Phạm Nhật Vượng was the first person from that country to be included in this list.\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Carlos Slim & family|$73.0 billion|73|Mexico|América Móvil, Grupo Carso|\n|2|Bill Gates|$67.0 billion|57|United States|Microsoft|\n|3|Amancio Ortega|$57.0 billion|76|Spain|Inditex Group|\n|4|Warren Buffett|$53.5 billion|82|United States|Berkshire Hathaway|\n|5|Larry Ellison|$43.0 billion|68|United States|Oracle Corporation|\n|6|Charles Koch|$34.0 billion|77|United States|Koch Industries|\n|6|David Koch|$34.0 billion|72|United States|Koch Industries|\n|8|Li Ka-shing|$31.0 billion|84|Hong Kong|Cheung Kong Holdings|\n|9|Liliane Bettencourt & family|$30.0 billion|90|France|L\'Oréal|\n|10|Bernard Arnault|$29.0 billion|63|France|LVMH|\n\n2012\n\nCarlos Slim topped the 2012 list, marking his third consecutive year at the top. Gates placed second\nbut narrowed the gap from 2011 as Slim\'s fortune fell $5 billion while Gates\' rose $5 billion. Warren\nBuffett remained in third place. Bernard Arnault of France was the top-ranking European on the list,\nplacing fourth.[34] Ricardo Salinas Pliego was the greatest gainer in terms of dollars, adding $9.2\nbillion to his fortune and moving up to number 37 overall.[34] Making her debut on the list at age 27,\nSpanx founder Sara Blakely became the youngest self-made female billionaire ever.[35] Colombia\'s\nAlejandro Santo Domingo was the highest-ranked newcomer, inheriting a $9.5 billion stake in Santo\nDomingo Group from his father.[34] India\'s Lakshmi Mittal was the largest loser as his fortune\ndropped from $31.1 billion to $20.7 billion as the price of steelmaker ArcelorMittal fell sharply.[34] As\na result, he failed to make the top 10 for the first time since 2004 and lost his title of richest Asian to\nHong Kong\'s Li Ka-shing.[36]\n\nA record total of 1,226 people made the 2012 list, representing 58 countries. Of those, 126 were\nnewcomers to the list and 104 were women. The United States had the greatest number of billionaires\nwith 425. Russia had 96 people on the list, while China had 95.[34] Georgia, Morocco, and Peru were\nnewly represented on the list. Falling stock prices in Asia contributed to 117 former billionaires falling\nfrom the list worldwide. Twelve others listed in 2011 died. Overall, net gainers (460) barely\noutnumbered net losers (441).[36]\n\nTo coincide with the release of the 2012 list, Forbes announced a then-new "Billionaire Real-Time\nTicker" updating the wealth of the world\'s top 50 billionaires in real time.[34]\n\nSource: https://en.wikipedia.org/wiki/The_World%27s_Billionaires\n---\n# The World\'s Billionaires - Wikipedia\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Carlos Slim & family|$69.0 billion|72|Mexico|América Móvil, Grupo Carso|\n|2|Bill Gates|$61.0 billion|56|United States|Microsoft|\n|3|Warren Buffett|$44.0 billion|81|United States|Berkshire Hathaway|\n|4|Bernard Arnault|$41.0 billion|63|France|LVMH Moët Hennessy • Louis Vuitton|\n|5|Amancio Ortega|$37.5 billion|75|Spain|Inditex Group|\n|6|Larry Ellison|$36.0 billion|67|United States|Oracle Corporation|\n|7|Eike Batista|$30.0 billion|55|Brazil|EBX Group|\n|8|Stefan Persson|$26.0 billion|64|Sweden|H&M|\n|9|Li Ka-shing|$25.5 billion|83|Hong Kong|Cheung Kong Holdings|\n|10|Karl Albrecht|$25.4 billion|92|Germany|Aldi|\n\nIn the 25th annual Forbes list of global billionaires, Slim added $20.5 billion to his fortune, the most of anyone, and retained his number one ranking with a total fortune of $74 billion. Gates remained in second place with $56 billion, while Warren Buffett was third with $50 billion. The top 10 had a combined wealth of $406 billion, up from $342 billion in 2010. According to Forbes editor Kerry Dolan, "media and technology billionaires definitely benefited from a stronger stock market and a growing enthusiasm for all things social" since the 2010 list. However, Nigerian commodity mogul Aliko Dangote was the greatest gainer on a percentage basis as his fortune increased 557 percent to $13.5 billion. Mark Zuckerberg was one of seven Facebook-related billionaires on the list, as he added $9.5 billion to his net worth to move up to 52nd. Facebook co-founder Dustin Moskovitz was the youngest person on the list. Aged 26, eight days younger than Zuckerberg, he debuted at number 420 with an estimated fortune of $2.7 billion. IKEA founder Ingvar Kamprad was the largest loser as he saw his fortune plummet from $23 billion to $6 billion, dropping him from 11th to 162nd overall.\n\nA record 1,210 billionaires made the 2011 list, representing a combined wealth of $4.5 trillion, up from $3.6 trillion the previous year. One third of the world\'s billionaires, 413, came from the United States. China had the second most billionaires with 115, while Russia was third with 101. Asia moved up to 332 billionaires, passing Europe as a region for the first time since the 1990s. The 2011 list included 214 newcomers and the average net worth of those on it increased to $3.7 billion.\n---\n# The World\'s Billionaires - Wikipedia\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Carlos Slim|$74.0 billion|71|Mexico|América Móvil, Grupo Carso|\n|2|Bill Gates|$56.0 billion|55|United States|Microsoft|\n|3|Warren Buffett|$50.0 billion|80|United States|Berkshire Hathaway|\n|4|Bernard Arnault|$41.0 billion|62|France|LVMH Moët Hennessy • Louis Vuitton|\n|5|Larry Ellison|$39.5 billion|66|United States|Oracle Corporation|\n|6|Lakshmi Mittal|$31.1 billion|60|India|Arcelor Mittal|\n|7|Amancio Ortega|$31.0 billion|74|Spain|Inditex Group|\n|8|Eike Batista|$30.0 billion|53|Brazil|EBX Group|\n|9|Mukesh Ambani|$27.0 billion|54|India|Reliance Industries|\n|10|Christy Walton & family|$26.5 billion|62|United States|Walmart|\n\nSlim narrowly eclipsed Gates to top the billionaire list for the first time. Slim saw his estimated worth surge $18.5 billion to $53.5 billion as shares of America Movil rose 35 percent. Gates\' estimated wealth rose $13 billion to $53 billion, placing him second. Warren Buffett was third with $47 billion. Christy Walton was the highest-ranking woman, placing 12th overall, with an inherited fortune of $22.5 billion. At age 25, Mark Zuckerberg continued to be the world\'s youngest self-made billionaire. American Isaac Perlmutter was among the newcomers with an estimated fortune of $4 billion largely acquired in his sale of Marvel Entertainment to Disney.\n\nA total of 1,011 people made the 2010 list. The United States accounted for 403 billionaires, followed by China with 89 and Russia with 62. It was the first time China, while including Hong Kong, placed second. A total of 55 countries were represented on the 2010 list, including Finland and Pakistan which claimed their first billionaires. Eighty-nine women made the list, but only 14 of them were self-made. The combined net worth of the list was $3.6 trillion, up 50 percent from 2009\'s $2.4 trillion, while the average net worth was $3.5 billion. The 2010 list featured 164 re-entries and 97 true newcomers. Asia accounted for more than 100 of the new entrants. Overall, just 12 percent of the list lost wealth since 2009, and 30 people fell off the list. 13 others died. Of the 89 women, 12 were newcomers in 2010. Steve Forbes said the growing number of billionaires was a clear sign that the world\'s economy was recovering from 2009\'s global financial crisis.\n\nIn June 2010, Gates and Buffett announced the Giving Pledge, a promise to give the majority of their wealth to philanthropic causes. As of 2017, the pledge had 158 signatories, but some of the signatories have since died. Most of the signers of the pledge are billionaires, and their pledges total over $365 billion.\n---\n# The World\'s Billionaires - Wikipedia\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Carlos Slim & family|$53.5 billion|70|Mexico|América Móvil, Grupo Carso|\n|2|Bill Gates|$53.0 billion|54|United States|Microsoft|\n|3|Warren Buffett|$47.0 billion|79|United States|Berkshire Hathaway|\n|4|Mukesh Ambani|$29.0 billion|53|India|Reliance Industries|\n|5|Lakshmi Mittal|$28.7 billion|60|India|Arcelor Mittal|\n|6|Larry Ellison|$28.0 billion|66|United States|Oracle Corporation|\n|7|Bernard Arnault|$27.5 billion|61|France|LVMH Moët Hennessy • Louis Vuitton|\n|8|Eike Batista|$27.0 billion|53|Brazil|EBX Group|\n|9|Amancio Ortega|$25.0 billion|74|Spain|Inditex Group|\n|10|Karl Albrecht|$23.5 billion|90|Germany|Aldi Süd|\n\nIn the wake of the Financial crisis of 2007–2008, the world\'s billionaires lost $2 trillion in net worth and the list became 30% smaller than the previous year\'s list.\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$40.0 billion|53|United States|Microsoft|\n|2|Warren Buffett|$37.0 billion|78|United States|Berkshire Hathaway|\n|3|Carlos Slim|$35.0 billion|69|Mexico|América Móvil, Grupo Carso|\n|4|Larry Ellison|$22.5 billion|64|United States|Oracle Corporation|\n|5|Ingvar Kamprad|$22.0 billion|83|Sweden|IKEA|\n|6|Karl Albrecht|$21.5 billion|89|Germany|Aldi Süd|\n|7|Mukesh Ambani|$19.5 billion|52|India|Reliance Industries|\n|8|Lakshmi Mittal|$19.3 billion|58|India|Arcelor Mittal|\n|9|Theo Albrecht|$18.8 billion|87|Germany|Aldi Nord, Trader Joe\'s|\n|10|Amancio Ortega|$18.3 billion|73|Spain|Inditex Group|\n\nFacebook founder Mark Zuckerberg, four years after starting the company, joined the list at 23 to become the youngest self-made billionaire.\n---\n# The World\'s Billionaires - Wikipedia\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Warren Buffett|$62.0 billion|77|United States|Berkshire Hathaway|\n|2|Carlos Slim|$60.0 billion|68|Mexico|América Móvil, Grupo Carso|\n|3|Bill Gates|$58.0 billion|52|United States|Microsoft|\n|4|Lakshmi Mittal|$45.0 billion|57|India|Arcelor Mittal|\n|5|Mukesh Ambani|$43.0 billion|51|India|Reliance Industries|\n|6|Anil Ambani|$42.0 billion|48|India|Anil Dhirubhai Ambani Group|\n|7|Ingvar Kamprad|$31.0 billion|81|Sweden|IKEA|\n|8|Kushal Pal Singh|$30.0 billion|76|India|DLF Group|\n|9|Oleg Deripaska|$28.0 billion|40|Russia|Rusal|\n|10|Karl Albrecht|$27.0 billion|88|Germany|Aldi Süd|\n\n2007\n\nForbes recorded a then record of 946 billionaires. There were 178 newcomers, as well as the first billionaires from Cyprus, Oman, Romania and Serbia. Over 66% of the previous year\'s billionaires became richer. The billionaires\' net worth increased in 2007 by $900 billion to $3.5 trillion.\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$56.0 billion|51|United States|Microsoft|\n|2|Warren Buffett|$52.0 billion|76|United States|Berkshire Hathaway|\n|3|Carlos Slim|$49.0 billion|67|Mexico|América Móvil, Grupo Carso|\n|4|Ingvar Kamprad|$33.0 billion|80|Sweden|IKEA|\n|5|Lakshmi Mittal|$32.0 billion|56|India|Arcelor Mittal|\n|6|Sheldon Adelson|$26.5 billion|73|United States|Las Vegas Sands|\n|7|Bernard Arnault|$26.0 billion|58|France|LVMH|\n|8|Amancio Ortega|$24.0 billion|71|Spain|Inditex Group|\n|9|Li Ka-shing|$23.0 billion|78|Hong Kong|Cheung Kong Holdings, Hutchison Whampoa|\n|10|David Thomson|$22.0 billion|49|Canada|Thomson Corporation|\n\n2006\n\nFree cash used by consumers from home equity extraction, known as the real estate bubble created a total of nearly $5 trillion in 2005, contributing to economic growth worldwide.\n\nSource\n---\n# The World\'s Billionaires - Wikipedia\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$52.0 billion|50|United States|Microsoft|\n|2|Warren Buffett|$42.0 billion|75|United States|Berkshire Hathaway|\n|3|Carlos Slim|$30.0 billion|66|Mexico|América Móvil, Grupo Carso|\n|4|Ingvar Kamprad|$28.0 billion|79|Sweden|IKEA|\n|5|Lakshmi Mittal|$23.5 billion|55|India|Mittal Steel Company|\n|6|Paul Allen|$22.0 billion|53|United States|Microsoft|\n|7|Bernard Arnault|$21.5 billion|57|France|LVMH Moët Hennessy • Louis Vuitton|\n|8|Al-Waleed bin Talal|$20.0 billion|49|Saudi Arabia|Kingdom Holding Company|\n|9|Kenneth Thomson|$19.6 billion|82|Canada|Thomson Corporation|\n|10|Li Ka-shing|$18.8 billion|77|Hong Kong|Cheung Kong Group, Hutchison Whampoa|\n\nThe net worth of 2005\'s 691 billionaires was $2.2 trillion. More than half of them had self-made fortunes.\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$46.5 billion|49|United States|Microsoft|\n|2|Warren Buffett|$44.0 billion|74|United States|Berkshire Hathaway|\n|3|Lakshmi Mittal|$25.0 billion|54|India|Mittal Steel Company|\n|4|Carlos Slim|$23.8 billion|65|Mexico|América Móvil, Grupo Carso|\n|5|Al-Waleed bin Talal|$23.7 billion|49|Saudi Arabia|Kingdom Holding Company|\n|6|Ingvar Kamprad|$23.0 billion|79|Sweden|IKEA|\n|7|Paul Allen|$21.0 billion|52|United States|Microsoft|\n|8|Karl Albrecht|$18.5 billion|85|Germany|Aldi Süd|\n|9|Larry Ellison|$18.4 billion|60|United States|Oracle Corporation|\n|10|S. Robson Walton|$18.3 billion|61|United States|Walmart|\n\nThe founders of Google, Sergey Brin and Larry Page, became billionaires at age 30.\n---\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$46.6 billion|48|United States|Microsoft|\n|2|Warren Buffett|$42.9 billion|73|United States|Berkshire Hathaway|\n|3|Karl Albrecht|$23.0 billion|84|Germany|Aldi Süd|\n|4|Al-Waleed bin Talal|$21.5 billion|47|Saudi Arabia|Kingdom Holding Company|\n|5|Paul Allen|$21.0 billion|51|United States|Microsoft|\n|6|Alice Walton*|$20.0 billion|55|United States|Wal-Mart|\n|6|Helen Walton*|$20.0 billion|84|United States|Wal-Mart|\n|6|Jim Walton*|$20.0 billion|56|United States|Wal-Mart|\n|6|John Walton*|$20.0 billion|58|United States|Wal-Mart|\n|6|S. Robson Walton*|$20.0 billion|60|United States|Wal-Mart|\n\n* Each hold an essentially equal share in Walmart.\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$40.7 billion|47|United States|Microsoft|\n|2|Warren Buffett|$30.5 billion|72|United States|Berkshire Hathaway|\n|3|Karl and Theo Albrecht|$25.6 billion|83|Germany|Aldi Süd|\n|4|Paul Allen|$20.1 billion|50|United States|Microsoft|\n|5|Al-Waleed bin Talal|$17.7 billion|46|Saudi Arabia|Kingdom Holding Company|\n|6|Larry Ellison|$16.6 billion|58|United States|Oracle Corporation|\n|7|Alice Walton*|$16.5 billion|54|United States|Wal-Mart|\n|7|Helen Walton*|$16.5 billion|83|United States|Wal-Mart|\n|7|Jim Walton*|$16.5 billion|55|United States|Wal-Mart|\n|7|John Walton*|$16.5 billion|57|United States|Wal-Mart|\n|7|S. Robson Walton*|$16.5 billion|59|United States|Wal-Mart|\n\n* Each hold an essentially equal share in Walmart.\n\n2003\n\nOprah Winfrey became the first female African-American billionaire.\n\n2002\n\nAs a result of the market crash caused by the Dot-com bubble, 83 billionaires dropped off the list from the previous year.\n---\n# The World\'s Billionaires - Wikipedia\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$52.8 billion|46|United States|Microsoft|\n|2|Warren Buffett|$35.0 billion|71|United States|Berkshire Hathaway|\n|3|Karl and Theo Albrecht|$26.8 billion|82|Germany|Aldi Süd|\n|4|Paul Allen|$25.2 billion|49|United States|Microsoft|\n|5|Larry Ellison|$23.5 billion|57|United States|Oracle Corporation|\n|6|Jim Walton*|$20.8 billion|54|United States|Wal-Mart|\n|7|John Walton*|$20.7 billion|56|United States|Wal-Mart|\n|8|Alice Walton*|$20.5 billion|53|United States|Wal-Mart|\n|8|S. Robson Walton*|$20.5 billion|58|United States|Wal-Mart|\n|8|Helen Walton*|$20.5 billion|82|United States|Wal-Mart|\n\n* Each hold an essentially equal share in Walmart.\n\nIn 2001, BET founder Robert L. Johnson became the first ever African-American billionaire.\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$58.7 billion|45|United States|Microsoft|\n|2|Warren Buffett|$32.3 billion|70|United States|Berkshire Hathaway|\n|3|Paul Allen|$30.4 billion|48|United States|Microsoft|\n|4|Larry Ellison|$26.0 billion|56|United States|Oracle Corporation|\n|5|Karl and Theo Albrecht|$25.0 billion|81|Germany|Aldi|\n|6|Al-Waleed bin Talal|$20.0 billion|44|Saudi Arabia|Kingdom Holding Company|\n|7|Jim Walton*|$18.8 billion|53|United States|Wal-Mart|\n|8|John Walton*|$18.7 billion|55|United States|Wal-Mart|\n|9|S. Robson Walton*|$18.6 billion|57|United States|Wal-Mart|\n|10|Alice Walton*|$18.5 billion|52|United States|Wal-Mart|\n|10|Helen Walton*|$18.5 billion|81|United States|Wal-Mart|\n\n* Each hold an essentially equal share in Wal-Mart. Had he been alive in 2001, Sam Walton would have been the world\'s wealthiest person.\n\nGates became the first American to take the top spot of the world\'s billionaires in 1995 with a net worth of $12.5 billion, and he remained there during the dot-com bubble\'s height in 1999 when his fortune peaked at $90 billion. After the dot-com bubble started to collapse in 2000, his wealth dropped to $60 billion, although he remained at the top of the list.\n---\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$60.0 billion|44|United States|Microsoft|\n|2|Larry Ellison|$47.0 billion|55|United States|Oracle Corporation|\n|3|Paul Allen|$28.0 billion|47|United States|Microsoft|\n|4|Warren Buffett|$25.6 billion|69|United States|Berkshire Hathaway|\n|5|Karl and Theo Albrecht|$20.0 billion|80|Germany|Aldi Süd|\n|6|Al-Waleed bin Talal|$20.0 billion|43|Saudi Arabia|Kingdom Holding Company|\n|7|S. Robson Walton|$20.0 billion|57|United States|Wal-Mart|\n|8|Masayoshi Son|$19.4 billion|43|Japan|Softbank Capital, SoftBank Mobile|\n|9|Michael Dell|$19.1 billion|35|United States|Dell|\n|10|Kenneth Thomson|$16.1 billion|77|Canada|The Thomson Corporation|\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$90.0 billion|43|United States|Microsoft|\n|2|Warren Buffett|$36.0 billion|68|United States|Berkshire Hathaway|\n|3|Paul Allen|$30.0 billion|46|United States|Microsoft|\n|4|Steven Ballmer|$19.5 billion|43|United States|Microsoft|\n|5|Philip Anschutz|$16.5 billion|59|United States|The Anschutz Corporation|\n|6|Michael Dell|$16.5 billion|34|United States|Dell|\n|7|S. Robson Walton|$15.8 billion|55|United States|Wal-Mart|\n|8|Al-Waleed Bin Talal|$15.0 billion|42|Saudi Arabia|Kingdom Holding Company|\n|9|Karl and Theo Albrecht|$13.6 billion|79|Germany|Aldi Süd|\n|10|Li Ka-shing & family|$12.6 billion|71|Hong Kong|CK Asset Holdings[48]|\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$51.0 billion|43|United States|Microsoft|\n---\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|2|Walton family|$48.0 billion|_|United States|Wal-Mart|\n|3|Warren Buffett|$33.0 billion|67|United States|Berkshire Hathaway|\n|4|Paul Allen|$21.0 billion|45|United States|Microsoft|\n|5|Kenneth Thomson|$14.4 billion|74|Canada|Woodbridge Co. Ltd.|\n|6|Jay and Robert Pritzker|$13.5 billion|_|United States|Hyatt|\n|7|Forrest Mars Sr. & family|$13.5 billion|94|United States|Mars, Inc.|\n|8|Al-Waleed Bin Talal|$13.3 billion|41|Saudi Arabia|Kingdom Holding Company|\n|9|Lee Shau-kee|$12.7 billion|70|Hong Kong|Henderson Land Development|\n|10|Karl and Theo Albrecht|$11.7 billion|78|Germany|Aldi Süd|\n\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Bill Gates|$36.4 billion|42|United States|Microsoft|\n|2|Walton family|$27.6 billion|_|United States|Wal-Mart|\n|3|Warren Buffett|$23.2 billion|66|United States|Berkshire Hathaway|\n|4|Lee Shau-kee|$14.7 billion|69|Hong Kong|Henderson Land Development|\n|5|Paul Allen|$14.1 billion|44|United States|Microsoft|\n|6|Kwok brothers|$12.3 billion|48|Hong Kong|Sun Hung Kai Properties|\n|7|Haas family|$12.3 billion|_|United States|Levi Strauss & Co|\n|8|Forrest Mars Sr. & family|$12.0 billion|93|United States|Mars, Inc.|\n|9|Karl and Theo Albrecht|$11.5 billion|77|Germany|Aldi Süd|\n|10|Tsai Wan-lin & family|$11.3 billion|73|Taiwan|Cathay Life Insurance|\n---\n|No.|Name|Net worth (USD)|Age|Nationality|Source(s) of wealth|\n|---|---|---|---|---|---|\n|1|Walton family|$22.9 billion|_|United States|Wal-Mart|\n|2|Bill Gates|$18.0 billion|41|United States|Microsoft|\n|3|Warren Buffett|$15.3 billion|65|United States|Berkshire Hathaway|\n|4|Oeri, Hoffman & Sacher families|$13.1 billion|_|Switzerland|Roche[56]|\n|5|Lee Shau-kee|$12.7 billion|68|Hong Kong|Henderson Land Development[52]|\n|6|Tsai Wan-lin & family|$12.2 billion|72|Taiwan|Cathay Life Insurance[55]|\n|7|Kwok brothers|$11.2 billion|_|Hong Kong|Sun Hung Kai Properties[53]|\n|8|Li Ka-shing & family|$10.6 billion|68|Hong Kong|CK Asset Holdings[48]|\n|9|Yoshiaki Tsutsumi|$9.2 billion|62|Japan|Seibu Railway[57]|\n|10|Karl and Theo Albrecht|$9.0 billion|76|Germany|Aldi Süd[58]|\n\n|No.|Name|Net worth (USD)|Nationality|\n|---|---|---|---|\n|1|Walton family|$23.5 billion|United States|\n|2|Bill Gates|$12.9 billion|United States|\n|3|Warren Buffett|$10.7 billion|United States|\n|4|Hans and Gad Rausing|$9.0 billion|Sweden|\n|5|Yoshiaki Tsutsumi|$9.0 billion|Japan|\n|6|Paul Sacher & Hoffman family|$8.6 billion|Switzerland|\n|7|Tsai Wan-lin & family|$8.5 billion|Taiwan|\n|8|Kenneth Thomson|$6.5 billion|Canada|\n|9|Lee Shau-kee|$6.5 billion|Hong Kong|\n|10|Chung Ju-yung|$6.2 billion|South Korea|\n\n|No.|Name|Net worth (USD)|Nationality|\n|---|---|---|---|\n|1|Walton family|$22.6 billion|United States|\n|2|du Pont family|$9.0 billion|United States|\n|3|Hans and Gad Rausing|$9.0 billion|Sweden|\n---\n|No.[47]|Name|Net worth (USD)|Nationality|\n|---|---|---|---|\n|4|Yoshiaki Tsutsumi|$8.5 billion|Japan|\n|5|Bill Gates|$8.2 billion|United States|\n|6|Warren Buffett|$7.9 billion|United States|\n|7|Paul Sacher & Hoffman family|$7.8 billion|Switzerland|\n|8|Tsai Wan-lin & family|$7.5 billion|Taiwan|\n|9|Karl and Theo Albrecht|$7.3 billion|Germany|\n|10|Carlos Slim|$6.6 billion|Mexico|\n\n|No.[47]|Name|Net worth (USD)|Nationality|\n|---|---|---|---|\n|1|Walton family|$25.3 billion|United States|\n|2|Mars family|$9.2 billion|United States|\n|3|Yoshiaki Tsutsumi|$9.0 billion|Japan|\n|4|du Pont family|$8.6 billion|United States|\n|5|Minoru and Akira Mori|$7.5 billion|Japan|\n|6|Bill Gates|$7.4 billion|United States|\n|7|Samuel and Donald Newhouse|$7.0 billion|United States|\n|8|Sid and Lee Bass & brothers|$6.8 billion|United States|\n|9|Warren Buffett|$6.6 billion|United States|\n|10|Erivan Haub|$6.2 billion|Germany|\n\n|No.[47]|Name|Net worth (USD)|Nationality|\n|---|---|---|---|\n|1|Walton family|$23.8 billion|United States|\n|2|Taikichiro Mori|$13.0 billion|Japan|\n|3|Yoshiaki Tsutsumi|$10.0 billion|Japan|\n|4|Hans and Gad Rausing|$7.0 billion|Sweden|\n|5|Erivan Haub|$6.9 billion|Germany|\n|6|Haniel family|$6.4 billion|Germany|\n|7|Bill Gates|$6.4 billion|United States|\n|8|David Sainsbury & family|$6.2 billion|United Kingdom|\n|9|Kenneth Thomson|$6.2 billion|Canada|\n|10|Shin Kyuk-ho|$6.0 billion|South Korea|\n---\n|No.|Name|Net worth (USD)|Nationality|\n|---|---|---|---|\n|1|Walton family|$18.5 billion|United States|\n|2|Taikichiro Mori|$15.0 billion|Japan|\n|3|Yoshiaki Tsutsumi|$14.0 billion|Japan|\n|4|du Pont family|$10.0 billion|United States|\n|5|Hans and Gad Rausing|$9.0 billion|Sweden|\n|6|Kitaro Watanabe|$7.7 billion|Japan|\n|7|Paul Reichmann & brothers|$7.1 billion|Canada|\n|8|Kenneth Thomson|$6.8 billion|Canada|\n|9|Kenkichi Nakajima|$6.1 billion|Japan|\n|10|Shin Kyuk-ho|$6.0 billion|South Korea|\n\n|No.|Name|Net worth (USD)|Nationality|\n|---|---|---|---|\n|1|Yoshiaki Tsutsumi|$16.0 billion|Japan|\n|2|Taikichiro Mori|$14.6 billion|Japan|\n|3|Walton family|$13.5 billion|United States|\n|4|du Pont family|$10.0 billion|United States|\n|5|Hans and Gad Rausing|$9.6 billion|Sweden|\n|6|Kitaro Watanabe|$9.2 billion|Japan|\n|7|Paul Reichmann & brothers|$9.0 billion|Canada|\n|8|Kenkichi Nakajima|$8.4 billion|Japan|\n|9|Shin Kyuk-ho|$7.5 billion|South Korea|\n|10|Eitaro Itoyama|$5.8 billion|Japan|\n\n|No.|Name|Net worth (USD)|Nationality|\n|---|---|---|---|\n|1|Yoshiaki Tsutsumi|$15.0 billion|Japan|\n|2|Taikichiro Mori|$14.2 billion|Japan|\n|3|Sam Walton & family|$8.7 billion|United States|\n|4|Reichmann brothers|$8.0 billion|Canada|\n---\n|No.|Name|Net worth (USD)|Nationality|\n|---|---|---|---|\n|4|Shin Kyuk-ho|$8.0 billion|South Korea|\n|6|Hirotomo Takei & family|$7.8 billion|Japan|\n|7|Kitaro Watanabe|$7.0 billion+|Japan|\n|8|Haruhiko Yoshimoto & family|$7.0 billion|Japan|\n|8|Hans and Gad Rausing|$7.0 billion|Sweden|\n|10|Eitaro Itoyama|$6.6 billion|Japan|\n\n|No.|Name|Net worth (USD)|Nationality|\n|---|---|---|---|\n|1|Yoshiaki Tsutsumi|$18.9 billion|Japan|\n|2|Taikichiro Mori|$18.0 billion|Japan|\n|3|Reichmann brothers|$9.0 billion|Canada|\n|4|Shin Kyuk-ho|$8.0 billion|South Korea|\n|4|K. C. Irving|$8.0 billion|Canada|\n|6|Haruhiko Yoshimoto|$7.8 billion|Japan|\n|7|Sam Walton|$6.5 billion|United States|\n|8|Tsai Wan-lin|$5.6 billion|Taiwan|\n|9|Eitaro Itoyama|$5.0 billion+|Japan|\n|10|Kitaro Watanabe|$5.2 billion|Japan|\n\n|No.|Name|Net worth (USD)|Nationality|\n|---|---|---|---|\n|1|Yoshiaki Tsutsumi|$20 billion|Japan|\n|2|Taikichiro Mori|$15 billion|Japan|\n|3|Shigeru Kobayashi|$7.5 billion|Japan|\n|4|Haruhiko Yoshimoto|$7.0 billion|Japan|\n|5|Salim Ahmed Bin Mahfouz|$6.2 billion|Saudi Arabia|\n|6|Hans and Gad Rausing|$6.0 billion|Sweden|\n|7|Paul Reichmann|$6.0 billion|Canada|\n|8|Yohachiro Iwasaki|$5.6 billion|Japan|\n|9|Kenneth Thomson|$5.4 billion|Canada|\n|10|Keizo Saji|$4.0 billion|Japan|\n---\nThe dot-com bubble created the most paper wealth for some billionaires. However, once the dot-com bubble burst the new rich saw their fortunes disappear. Billionaires\' fortunes were hit even harder by the global financial crisis; 2009 was the first time in five years that the world had a net loss in the number of billionaires. The strong performance of the financial markets and global economic recovery have erased financial assets losses. Most of the richest people in the world saw their fortunes soar in the early 2010s.\n\nSource\n---\n# Number and combined net worth of billionaires by year\n\n|Year|Number of billionaires|Group\'s combined net worth|\n|---|---|---|\n|2023[2]|2,640|$12.2 trillion|\n|2022[6]|2,668|$12.7 trillion|\n|2021[11]|2,755|$13.1 trillion|\n|2020|2,095|$8.0 trillion|\n|2019|2,153|$8.7 trillion|\n|2018|2,208|$9.1 trillion|\n|2017|2,043|$7.7 trillion|\n|2016|1,810|$6.5 trillion|\n|2015[18]|1,826|$7.1 trillion|\n|2014[67]|1,645|$6.4 trillion|\n|2013[68]|1,426|$5.4 trillion|\n|2012|1,226|$4.6 trillion|\n|2011|1,210|$4.5 trillion|\n|2010|1,011|$3.6 trillion|\n|2009|793|$2.4 trillion|\n|2008|1,125|$4.4 trillion|\n|2007|946|$3.5 trillion|\n|2006|793|$2.6 trillion|\n|2005|691|$2.2 trillion|\n|2004|587|$1.9 trillion|\n|2003|476|$1.4 trillion|\n|2002|497|$1.5 trillion|\n|2001|538|$1.8 trillion|\n|2000|470|$898 billion|\n\nSources: Forbes.[18][67][66][68]\n\n# See also\n\n- Biography portal\n- Economics portal\n- World portal\n- Bloomberg Billionaires Index\n- List of centibillionaires\n- List of cities by number of billionaires\n\nhttps://en.wikipedia.org/wiki/The_World%27s_Billionaires\n---\n# The World\'s Billionaires - Wikipedia\n\nList of countries by number of billionaires\n\nList of wealthiest animals\n\nList of wealthiest families\n\nList of wealthiest historical figures\n\nList of wealthiest religious organizations\n\n# References\n\n# Specific\n\n1. Kollmeyer, Barbara (4 March 2014). "Bill Gates back on top, WhatsApp founders and more\nwomen in Forbes\' billionaire rankings" (source). Market Watch, The Wall Street Journal. Archived from the original\n(source) on 31 August 2017. Retrieved 5 March 2014.\n2. Peterson-Withorn, Chase (4 April 2023). "Forbes\' 37th Annual World\'s Billionaires List: Facts And\nFigures 2023" (source). Forbes. Archived (source) from the original on 5 April 2023. Retrieved 5 April 2023.\n3. Gore, Leada (4 March 2014). "Youngest, oldest and one who can body slam you: More on Forbes\nlist of the ultra-rich" (source). AL.com. Alabama Media Group. Archived (source) from\nthe original on 28 June 2018. Retrieved 5 March 2014.\n4. "Why Putin Isn\'t on \'Forbes\' Billionaires List" (source). Forbes. 2 March 2015. Archived (source) from the original on 20\nOctober 2017. Retrieved 17 November 2017.\n5. Kroll, Luisa. "Forbes 2017 billionaires list: Meet the richest people on the planet" (source). Forbes. Archived (source) from the original on 21 March 2017. Retrieved 21 March 2017.\n6. Peterson-Withorn, Chase (5 April 2022). "Forbes\' 36th Annual World\'s Billionaires List: Facts And\nFigures 2022" (source). Forbes. Archived (source) from the original on 8 April 2022. Retrieved 8 April 2022.\n7. Dolan, Kerry A. (7 March 2012). "Methodology: How We Crunch The Numbers" (source). Forbes. Archived\n(source) from the original on 3 January 2013. Retrieved\n4 January 2013.\n---\n# The World\'s Billionaires - Wikipedia\n\n8. Krastev, Nikola (11 March 2010). "Forbes\' Rich List: Number Of New Billionaires Reflects Global Recovery" http://www.rferl.org/content/Forbes_Rich_List_Number_Of_New_Billionaires_Reflects_Global_Recovery/1980413.html. Radio Free Europe Radio Liberty. Archived https://web.archive.org/web/20140326024704/http://www.rferl.org/content/Forbes_Rich_List_Number_Of_New_Billionaires_Reflects_Global_Recovery/1980413.html from pe original on 26 March 2014. Retrieved 25 March 2014.\n9. Fortunes Of Kings, Queens And Dictators https://www.forbes.com/2006/05/04/rich-kings-dictators_cz_lk_0504royals.html Archived https://web.archive.org/web/20110219084918/http://www.forbes.com/2006/05/04/rich-kings-dictators_cz_lk_0504royals.html 19 February 2011 at pe Wayback Machine Forbes, 5 May 2006\n10. Liu, Phoebe (4 April 2023). "The World\'s Youngest Billionaires 2023" https://www.forbes.com/sites/phoebeliu/2023/04/04/pe-worlds-youngest-billionaires-2023/?sh=2dd4fdce7575. Forbes. Retrieved 6 April 2023.\n11. Dolan, Kerry A. (6 April 2021). "Forbes World\'s Billionaires List - The Richest in 2021" https://www.forbes.com/billionaires/. www.forbes.com. Archived https://web.archive.org/web/20220217164331/https://www.forbes.com/billionaires/ from pe original on 17 February 2022. Retrieved 18 February 2022.\n12. Dolan, Kerry A. "Forbes\' 35p Annual World\'s Billionaires List: Facts And Figures 2021" https://www.forbes.com/sites/kerryadolan/2021/04/06/forbes-35p-annual-worlds-billionaires-list-facts-and-figures-2021/. Forbes. Archived https://web.archive.org/web/20210409230211/https://www.forbes.com/sites/kerryadolan/2021/04/06/forbes-35p-annual-worlds-billionaires-list-facts-and-figures-2021/ from pe original on 9 April 2021. Retrieved 10 April 2021.\n13. Dolan, Kerry A. (2020). "Forbes Billionaires 2020" https://www.forbes.com/billionaires/. Forbes. Archived https://web.archive.org/web/20190104180124/https://www.forbes.com/billionaires/list/3/ from pe original on 4 January 2019. Retrieved 7 April 2020.\n14. "Billionaires 2019" https://web.archive.org/web/20190701053757/https://www.forbes.com/billionaires/. Forbes. 5 March 2019. Archived from pe original https://www.forbes.com/billionaires/#698401a8251c on 1 July 2019. Retrieved 13 March 2019.\n15. Kroll, Luisa (6 March 2018). "Forbes Billionaires 2018: Meet The Richest People On The Planet" https://www.forbes.com/sites/luisakroll/2018/03/06/forbes-billionaires-2018-meet-pe-richest-people-on-pe-planet/. Forbes. Archived https://web.archive.org/web/20180308165924/https://www.forbes.com/sites/luisakroll/2018/03/06/forbes-billionaires-2018-meet-pe-richest-people-on-pe-planet/ from pe original on 8 March 2018. Retrieved 6 March 2018.\n16. Dolan, Kerry A. "Why No Saudi Arabians Made The Forbes Billionaires List This Year" https://www.forbes.com/sites/kerryadolan/2018/03/06/no-saudi-arabian-billionaires-forbes-list-2018-alwaleed-alamoudi/. Forbes. Archived https://web.archive.org/web/20180306150806/https://www.forbes.com/sites/kerryadolan/2018/03/06/no-saudi-arabian-billionaires-forbes-list-2018-alwaleed-alamoudi/ from pe original on 6 March 2018. Retrieved 21 October 2019.\n17. "The World\'s Billionaires List" https://www.forbes.com/billionaires/list/#version:static_tab:women. Forbes. 23 March 2017. Archived https://web.archive.org/web/20190304152923/https://www.forbes.com/billionaires/list/#version:static_tab:women from pe original on 4 March 2019. Retrieved 21 March 2017.\n18. Dolan, Kerry A.; Kroll, Luisa (2 March 2015). "Inside The 2015 Forbes Billionaires List: Facts And Figures" https://www.forbes.com/sites/kerryadolan/2015/03/02/inside-pe-2015-forbes-billionaires-list-facts-and-figures/. Forbes. Archived https://web.archive.org/web/20180628205305/https://www.forbes.com/sites/kerryadolan/2015/03/02/inside-pe-2015-forbes-billionaires-list-facts-and-figures/ from pe original on 28 June 2018. Retrieved 31 May 2015.\n19. "#1 Bill Gates" https://web.archive.org/web/20150304235759/http://www.forbes.com/profile/bill-gates/?list=billionaires. Forbes. Archived from pe original https://www.forbes.com/profile/bill-gates/?list=billionaires on 4 March 2015. Retrieved 2 January 2016.\n\nhttps://en.wikipedia.org/wiki/The_World%27s_Billionaires\n---\n# The World\'s Billionaires - Wikipedia\n\n20. "Amancio Ortega" (https://www.forbes.com/profile/amancio-ortega/?list=billionaires). Forbes. 27 January 2016. Archived (https://web.archive.org/web/20190903053752/https://www.forbes.com/profile/amancio-ortega/?list=billionaires) from pe original on 3 September 2019. Retrieved 27 January 2016.\n21. "The World\'s Billionaires : Warren Buffett" (https://www.forbes.com/profile/warren-buffett/). Forbes. Archived (https://web.archive.org/web/20200815130703/https://www.forbes.com/profile/warren-buffett/) from pe original on 15 August 2020. Retrieved 3 November 2015.\n22. "Jeff Bezos" (https://www.forbes.com/profile/jeff-bezos/). Forbes. Archived (https://archive.today/2021.01.30-021706/https://www.forbes.com/profile/jeff-bezos/?sh=3c149a3a1b23) from pe original on 30 January 2021. Retrieved 27 October 2015.\n23. "The World\'s Billionaires - Mark Zuckerberg" (https://www.forbes.com/profile/mark-zuckerberg/?list=billionaires). Forbes. 1 March 2016. Archived (https://web.archive.org/web/20190320122956/https://www.forbes.com/profile/mark-zuckerberg/?list=billionaires) from pe original on 20 March 2019. Retrieved 21 March 2016.\n24. "The World\'s Billionaires - Michael Bloomberg" (https://www.forbes.com/profile/michael-bloomberg/?list=billionaires). Forbes. 1 March 2016. Archived (https://web.archive.org/web/20171011033351/https://www.forbes.com/profile/michael-bloomberg/?list=billionaires) from pe original on 11 October 2017. Retrieved 21 March 2016.\n25. Alexander, Dan (2 March 2015). "Record 290 Newcomers Join Forbes Billionaires List, Including Michael Jordan" (https://www.forbes.com/sites/danalexander/2015/03/02/record-290-newcomers-join-forbes-billionaires-list-including-michael-jordan). Forbes. Archived (https://web.archive.org/web/20170921085316/https://www.forbes.com/sites/danalexander/2015/03/02/record-290-newcomers-join-forbes-billionaires-list-including-michael-jordan/) from pe original on 21 September 2017. Retrieved 16 September 2017.\n26. Dolan, Kerry A. "The World\'s Billionaires" (https://www.forbes.com/sites/kerryadolan/2015/03/02/inside-pe-2015-forbes-billionaires-list-facts-and-figures). Forbes. Archived (https://web.archive.org/web/20150303053520/http://www.forbes.com/sites/kerryadolan/2015/03/02/inside-pe-2015-forbes-billionaires-list-facts-and-figures/) from pe original on 3 March 2015. Retrieved 2 March 2015.\n27. "The World\'s Billionaires" (https://www.forbes.com/billionaires/list/#version:static). Forbes. 3 March 2015. Archived (https://web.archive.org/web/20190304152923/https://www.forbes.com/billionaires/list/#version:static) from pe original on 4 March 2019. Retrieved 2 March 2015.\n28. "Mario Lopez Estrada" (https://www.forbes.com/profile/mario-lopez-estrada/). Forbes: The World\'s Billionaires. Archived (https://web.archive.org/web/20150829055113/http://www.forbes.com/profile/mario-lopez-estrada/) from pe original on 29 August 2015. Retrieved 19 August 2015.\n29. Dolan, Kerry A.; Kroll, Luisa (3 March 2014). "Inside The 2014 Forbes Billionaires List: Facts And Figures" (https://www.forbes.com/sites/luisakroll/2014/03/03/inside-pe-2014-forbes-billionaires-list-facts-and-figures/). Forbes. Archived (https://web.archive.org/web/20140602005842/http://www.forbes.com/sites/luisakroll/2014/03/03/inside-pe-2014-forbes-billionaires-list-facts-and-figures/) from pe original on 2 June 2014. Retrieved 2 June 2014.\n30. Barber, Elizabep (4 March 2014). "Forbes\' richest people: number of billionaires up significantly" (http://www.csmonitor.com/USA/USA-Update/2014/0303/Forbes-richest-people-number-of-billionaires-up-significantly-video). Christian Science Monitor. Archived (https://web.archive.org/web/20140304010932/http://www.csmonitor.com/USA/USA-Update/2014/0303/Forbes-richest-people-number-of-billionaires-up-significantly-video) from pe original on 4 March 2014. Retrieved 4 March 2014.\n---\n# The World\'s Billionaires - Wikipedia\n\n|31.|Hickman, Martin (4 March 2013). "2013 Forbes Billionaires list: Record number of new entries appear on rich list, but Carlos Slim and Bill Gates still top the charts" (source). The Independent. Archived (source) from the original on 25 September 2015. Retrieved 5 March 2014.|\n|---|---|\n|32.|Kollmeyer, Barbara (4 March 2013). "Forbes billionaire list 2013: Mr. Slim wins it again" (source). Market Watch, Wall Street Journal. Archived from the original (source) on 6 March 2014. Retrieved 5 March 2014.|\n|33.|Reaney, Patricia (4 March 2013). "Factbox: Record number of billionaires make Forbes 2013 list" (source). Reuters. Archived (source) from the original on 2 January 2016. Retrieved 5 March 2014.|\n|34.|"Forbes rich list 2012: Carlos Slim tops rich list for third year" (source). The Telegraph. 8 March 2012. Archived (source) from the original on 10 January 2022. Retrieved 8 March 2014.|\n|35.|"Spanx creator Sara Blakely becomes youngest woman on Forbes Billionaires List" (source). Fox News Channel. 8 March 2012. Archived (source) from the original on 5 March 2016. Retrieved 8 March 2014.|\n|36.|Kroll, Luisa (7 March 2014). "Forbes World\'s Billionaires 2012" (source). Forbes. Archived (source) from the original on 8 March 2012. Retrieved 8 March 2014.|\n|37.|Georg Szalai (9 March 2011). "Steven Spielberg, Oprah Winfrey Drop in Forbes Billionaire Ranking" (source). The Hollywood Reporter. Archived (source) from the original on 20 May 2014. Retrieved 15 March 2014.|\n|38.|Herbert, Geoff (10 March 2011). "Forbes\' 2011 list of \'World\'s Billionaires\' is biggest ever" (source). Syracuse Post-Standard. Archived (source) from the original on 15 March 2014. Retrieved 15 March 2014.|\n|39.|Miller, Matthew; Kroll, Luisa (10 March 2010). "Bill Gates No Longer World\'s Richest Man" (source). Forbes. Archived (source) from the original on 26 September 2011. Retrieved 25 March 2014.|\n---\n# The World\'s Billionaires - Wikipedia\n\n|40.|"China ranks No 2 on Forbes billionaires list" (source)|China Daily. 11 March 2010. Archived (source) from the original on 18 March 2014. Retrieved 17 March 2014.|\n|---|---|---|\n|41.|"The Giving Pledge List Reaches $365 Billion" (source)|Insatiable Fox. Archived (source) from the original on 20 July 2018. Retrieved 23 April 2016.|\n|42.|Confer Thomas Philippon. "The future of the financial industry" (source)|Finance Department of the New York University Stern School of Business. Archived (source) from the original on 2 March 2017. Retrieved 5 April 2018.|\n|43.|The World\'s Richest People (source)|Archived (source) 5 January 2019 at the Wayback Machine Forbes, 8 March 2007|\n|44.|"Greenspan Kennedy Report – Table 2 – Sources and Uses of Equity Extracted from Homes" (source) (PDF)|Archived (source) (PDF) from the original on 11 July 2012. Retrieved 1 May 2010.|\n|45.|"The World\'s Richest people" (source)|Forbes. 10 March 2005. Archived from the original (source) on 28 April 2006. Retrieved 28 April 2006.|\n|46.|Miller, Matthew (6 May 2009). "The Wealthiest Black Americans" (source)|Forbes Magazine. Archived from the original (source) on 15 December 2019. Retrieved 21 September 2013.|\n|47.|Dolan, Kerry A., ed. (5 July 1999). "200 Global Billionaires". Forbes. Vol. 164, no. 1. pp. 153–155.|EBSCOhost 1966292 (source).|\n|48.|"How Li Ka Shing became richest man in Hong Kong" (source)|Business Insider. 20 August 2019. Archived (source) from the original on 1 December 2021. Retrieved 1 December 2021.|\n|49.|Thomson.com. Management (source)|Archived (source) 23 April 2022 at the Wayback Machine. Accessed 23 March 2006.|\n|50.|New York Times: "Jay Pritzker, Who Built Chain Of Hyatt Hotels, Is Dead at 76" By ANTHONY RAMIREZ (source)|Archived (source) 2 December 2021 at the Wayback Machine 25 January 1999|\n\nSource\n---\n# The World\'s Billionaires - Wikipedia\n\n|51.|Brenner, Joel Glenn (26 July 1992). "Life on Mars: The Mars family saga has all the classic elements" (source). The Independent. Archived (source) from the original on 18 July 2010. Retrieved 25 February 2011.|\n|---|---|\n|52.|"Forbes Profile: Lee Shau Kee" (source). Forbes. 11 June 2021. Archived (source) from the original on 2 December 2021. Retrieved 2 December 2021.|\n|53.|Ramzy, Austin (22 October 2018). "NY Times Article: Walter Kwok" (source). The New York Times. Archived (source) from the original on 2 January 2022. Retrieved 2 January 2022.|\n|54.|Kenneth N. Gilpin (22 September 1995). "Walter A. Haas Jr., 79, Leader Of Family Behind Levi Strauss" (source). The New York Times. p. B7. Archived (source) from the original on 27 October 2021. Retrieved 2 January 2022.|\n|55.|"Tsai Wan-lin" (source). Chicago Tribune. 29 September 2004. Retrieved 6 December 2014.|\n|56.|"Switzerland\'s Biggest Family Fortune Has Prospered for 124 Years" (source). Bloomberg.com. 6 August 2020.|\n|57.|Tsutsumi was the richest man on earth during 1987-94, says Forbes (source)|\n|58.|"The World\'s Billionaires (2012): No. 10 Karl Albrecht" (source). Forbes. 3 March 2010. Retrieved 20 February 2011.|\n|59.|"Japanese businessman tops Forbes billionaires list again" (source). Business Monday. The Burlington Free Press. Associated Press. 10 July 1989. p. 12. Archived (source) from the original on 6 October 2021. Retrieved 6 October 2021 – via Newspapers.com.|\n|60.|"Japanese rail executive tops billionaires list again" (source). Business. The Sun. Associated Press. 8 July 1988. p. 3E. Archived (source) from the original on 6 October 2021. Retrieved 6 October 2021 – via Newspapers.com.|\n|61.|Kilachand, Sean. "Forbes History: The Original 1987 List Of International Billionaires" (source). Forbes. Archived (source) from the original on 27 February 2020. Retrieved 17 October 2021.|\n|62.|"Special Report on The New Rich" (source). The Economist. 14 June 2001. Archived (source) from the original on 26 February 2015. Retrieved 21 March 2012.|\n\nSource\n---\n# The World\'s Billionaires - Wikipedia\n\n|63.|"Special Report on The Rich" (http://www.economist.com/node/13356686). The Economist. 4 April 2009. Archived (https://web.archive.org/web/20110112194744/http://www.economist.com/node/13356686) from the original on 12 January 2011. Retrieved 19 March 2012.|\n|---|---|\n|64.|"Global Wealth Report 2011" (https://web.archive.org/web/20150923212754/http://www.dasinvestment.com/fileadmin/images/pictures/0907/BCG_2011_0106_Global_Wealth_Report_client_version_May2011.pdf) (PDF). The Boston Consulting Group. 31 May 2011. Archived from the original (http://www.dasinvestment.com/fileadmin/images/pictures/0907/BCG_2011_0106_Global_Wealth_Report_client_version_May2011.pdf) (PDF) on 23 September 2015.|\n|65.|"World Wealth Report 2011" (https://web.archive.org/web/20111103030847/http://ml.com/media/114235.pdf) (PDF). Capgemini and Merrill Lynch. 22 June 2011. Archived from the original (http://www.ml.com/media/114235.pdf) (PDF) on 3 November 2011.|\n|66.|"Billionaires 25th Anniversary Timeline – Forbes" (https://www.forbes.com/special-report/2012/billionaires-25th-anniversary-timeline.html). Forbes. Archived (https://web.archive.org/web/20140131061616/http://www.forbes.com/special-report/2012/billionaires-25th-anniversary-timeline.html) from the original on 31 January 2014. Retrieved 31 May 2015.|\n|67.|Dolan, Kerry A.; Kroll, Luisa (3 March 2014). "Inside The 2014 Forbes Billionaires List: Facts And Figures" (https://www.forbes.com/sites/luisakroll/2014/03/03/inside-the-2014-forbes-billionaires-list-facts-and-figures/). Forbes. Archived (https://web.archive.org/web/20150524020855/http://www.forbes.com/sites/luisakroll/2014/03/03/inside-the-2014-forbes-billionaires-list-facts-and-figures/) from the original on 24 May 2015. Retrieved 31 May 2015.|\n|68.|Kroll, Luisa (4 March 2013). "Inside The 2013 Billionaires List: Facts and Figures" (https://www.forbes.com/sites/luisakroll/2013/03/04/inside-the-2013-billionaires-list-facts-and-figures/). Forbes. Archived (https://web.archive.org/web/20180713175059/https://www.forbes.com/sites/luisakroll/2013/03/04/inside-the-2013-billionaires-list-facts-and-figures/) from the original on 13 July 2018. Retrieved 31 May 2015.|\n\n# General\n\n- 2000–2010 Top 10s: Rogers, Simon (11 March 2010). "Forbes rich list: ten years of top tens" (https://www.theguardian.com/news/datablog/2010/mar/11/forbes-rich-list-top-ten-carlos-slim). The Guardian.\n- 2011 Top 10: "The World\'s Billionaires 2011" (https://www.forbes.com/lists/2011/10/billionaires_2011.html). Forbes.\n- 2012 Top 10: "Forbes rich list 2012: top 50" (https://www.telegraph.co.uk/finance/9130322/Forbes-rich-list-2012-top-50.html). The Telegraph. 8 March 2012. Archived (https://ghostarchive.org/archive/20220110/https://www.telegraph.co.uk/finance/9130322/Forbes-rich-list-2012-top-50.html) from the original on 10 January 2022.\n- 2013 Top 10: Durgy, Edwin (4 March 2013). "The World\'s Richest Billionaires: Full List Of The Top 500" (https://www.forbes.com/sites/edwindurgy/2013/03/04/the-worlds-richest-billionaires-full-list-of-the-top-500/). Forbes.\n- 2014 Top 10: "Forbes Releases 28th Annual World\'s Billionaires Issue" (https://www.forbes.com/sites/forbespr/2014/03/03/forbes-releases-28th-annual-worlds-billionaires-issue/). Forbes. 3 March 2014. Retrieved 3 March 2014.\n---\n#\n\n7/1/23, 11:31 PM The World\'s Billionaires - Wikipedia\n2015 Top 10: "Forbes\' 29th Annual World\'s Billionaires Issue" (https://www.forbes.com/sites/forbes\npr/2015/03/02/forbes-29th-annual-worlds-billionaires-issue/). Forbes. 2 March 2015. Retrieved\n12 March 2015.\n\n# External links\n\nForbes: "The World\'s Billionaires" (https://www.forbes.com/billionaires/)\nForbes: Real-Time Billionaires (https://www.forbes.com/real-time-billionaires/)\n\nRetrieved from "https://en.wikipedia.org/w/index.php?title=The_World%27s_Billionaires&amp;oldid=1162639101"\n\nhttps://en.wikipedia.org/wiki/The_World%27s_Billionaires 33/33'

-----------------------------------------------------------------------

In [7]:
import os 
from pathlib import Path
from llama_index.readers.file import FlatReader

from llama_index.core.text_splitter import SentenceSplitter
from llama_index.core.text_splitter import TokenTextSplitter
from llama_index.embeddings.huggingface import HuggingFaceEmbedding

In [11]:
file_path = os.path.join('..','..','Data', 'input', 'Create a fulfilment to authorize a service _ Healthanea Documentation.html')
file_path

'..\\..\\Data\\input\\Create a fulfilment to authorize a service _ Healthanea Documentation.html'

In [12]:
reader = FlatReader()
docs = reader.load_data(Path(file_path))
docs

[Document(id_='a48fe48e-c6de-4800-b7ee-f2793e5ef5ab', embedding=None, metadata={'filename': 'Create a fulfilment to authorize a service _ Healthanea Documentation.html', 'extension': '.html'}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=[], relationships={}, text='\ufeff<!DOCTYPE html>\n<!--[if IE]><![endif]-->\n<html>\n  \n  <head>\n    <meta charset="utf-8">\n    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">\n    <title>Create a fulfilment to authorize a service | Healthanea Documentation </title>\n    <meta name="viewport" content="width=device-width">\n    <meta name="title" content="Create a fulfilment to authorize a service | Healthanea Documentation ">\n    <meta name="generator" content="docfx 2.59.0.0">\n    \n    <link rel="shortcut icon" href="../../favicon.ico">\n    <link rel="stylesheet" href="../../styles/docfx.vendor.css">\n    <link rel="stylesheet" href="../../styles/docfx.css">\n    <link rel="stylesheet" href="../../styles/main.css">

In [1]:
from llama_index.readers.web import SimpleWebPageReader

from llama_index.readers.file import FlatReader #Comparacion entre los dos ?? Porque uno mejor que otro ??? 
from llama_index.readers.file import HTMLTagReader


import os 
from pathlib import Path
file_path = os.path.join('..','..','Data', 'input', 'Create a fulfilment to authorize a service _ Healthanea Documentation.html')
file_path

c:\Users\ebt1023\AppData\Local\Programs\Python\Python312\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


'..\\..\\Data\\input\\Create a fulfilment to authorize a service _ Healthanea Documentation.html'

In [18]:
HTMLTagReader(tag='article').load_data(Path(file_path))

[Document(id_='f01d4c31-54d5-4b94-ab87-c819a1c35244', embedding=None, metadata={'tag': 'article', 'tag_id': '_content', 'file_path': '..\\..\\Data\\input\\Create a fulfilment to authorize a service _ Healthanea Documentation.html'}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=[], relationships={}, text='Create a fulfilment to authorize a service\nSee also the API specification (Swagger) and Error management\n\nSee also:\nConcept: Service fulfilment\nAPI documentation: The Fulfilment API\nProducer Channel role: You use the Fulfilment API to create a fulfilment instance with which you authorize Service Providers to offer healthcare services to end-users.\nService role: You use the Fulfilment API to register actions related to your service\'s lifecycle within the context of the fulfilment instance. These can include actions performed by the end-user but also the read and write actions applied to the end-user\'s health data.\nThe Fulfilment API tracks the actions performed 

In [9]:
reader = FlatReader()
document = reader.load_data(Path(file_path))
document

[Document(id_='705d1dfc-4be7-463b-aa19-c56cab30a808', embedding=None, metadata={'filename': 'Create a fulfilment to authorize a service _ Healthanea Documentation.html', 'extension': '.html'}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=[], relationships={}, text='\ufeff<!DOCTYPE html>\n<!--[if IE]><![endif]-->\n<html>\n  \n  <head>\n    <meta charset="utf-8">\n    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">\n    <title>Create a fulfilment to authorize a service | Healthanea Documentation </title>\n    <meta name="viewport" content="width=device-width">\n    <meta name="title" content="Create a fulfilment to authorize a service | Healthanea Documentation ">\n    <meta name="generator" content="docfx 2.59.0.0">\n    \n    <link rel="shortcut icon" href="../../favicon.ico">\n    <link rel="stylesheet" href="../../styles/docfx.vendor.css">\n    <link rel="stylesheet" href="../../styles/docfx.css">\n    <link rel="stylesheet" href="../../styles/main.css">

In [23]:
document[0].text

'\ufeff<!DOCTYPE html>\n<!--[if IE]><![endif]-->\n<html>\n  \n  <head>\n    <meta charset="utf-8">\n    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">\n    <title>Create a fulfilment to authorize a service | Healthanea Documentation </title>\n    <meta name="viewport" content="width=device-width">\n    <meta name="title" content="Create a fulfilment to authorize a service | Healthanea Documentation ">\n    <meta name="generator" content="docfx 2.59.0.0">\n    \n    <link rel="shortcut icon" href="../../favicon.ico">\n    <link rel="stylesheet" href="../../styles/docfx.vendor.css">\n    <link rel="stylesheet" href="../../styles/docfx.css">\n    <link rel="stylesheet" href="../../styles/main.css">\n    <link rel="stylesheet" href="../../styles/dhp.css">\n    <link href="https://fonts.googleapis.com/css?family=Roboto" rel="stylesheet"> \n    <meta property="docfx:navrel" content="../../toc.html">\n    <meta property="docfx:tocrel" content="../toc.html">\n    \n    <meta pr

In [28]:
docs = SimpleWebPageReader(html_to_text = True).html_to_text(document[0].text)  ### SEARCH IN THE INTERNET OR CHATBOT :) 
docs

TypeError: 'bool' object is not callable

----------------------------------------------------------------------------

### General Test

In [1]:
import os 
from pathlib import Path
from llama_index.readers.file import FlatReader

from llama_index.core.text_splitter import SentenceSplitter
from llama_index.embeddings.huggingface import HuggingFaceEmbedding

file_path = os.path.join('..','..','Data', 'input', 'Create a fulfilment to authorize a service _ Healthanea Documentation.html')

reader = FlatReader()
docs = reader.load_data(Path(file_path))

nodes = SentenceSplitter( chunk_size = 700, chunk_overlap = 3).get_nodes_from_documents(docs)
nodes

c:\Users\ebt1023\AppData\Local\Programs\Python\Python312\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


[TextNode(id_='7e3445c3-0904-4e09-9c3c-9b0f48d6c037', embedding=None, metadata={'filename': 'Create a fulfilment to authorize a service _ Healthanea Documentation.html', 'extension': '.html'}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=[], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='a2a51a01-c3df-4b94-aa3b-1897daba7671', node_type=<ObjectType.DOCUMENT: '4'>, metadata={'filename': 'Create a fulfilment to authorize a service _ Healthanea Documentation.html', 'extension': '.html'}, hash='176daef87353e632b6b7e46bc77b6d8cf20f5ce1392401cc3694166a7935db26'), <NodeRelationship.NEXT: '3'>: RelatedNodeInfo(node_id='9a5ecca9-70bb-4606-982b-71328cef7699', node_type=<ObjectType.TEXT: '1'>, metadata={}, hash='5f4c7766105ce67d7d1078f9c190403a4f54348ab2ceb2c65dc7911e88051ae3')}, text='\ufeff<!DOCTYPE html>\n<!--[if IE]><![endif]-->\n<html>\n  \n  <head>\n    <meta charset="utf-8">\n    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">\n    <tit

In [2]:
os.environ['GROQ_API_KEY'] = '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'

In [41]:
from llama_index.embeddings.huggingface import HuggingFaceEmbedding
from llama_index.embeddings.openai import OpenAIEmbedding
from llama_index.llms.groq import Groq
from llama_index.core import ServiceContext, StorageContext, VectorStoreIndex

from llama_index.core.retrievers import QueryFusionRetriever
from llama_index.retrievers.bm25 import BM25Retriever

from llama_index.core.postprocessor import LLMRerank
from llama_index.postprocessor.jinaai_rerank import JinaRerank 
from llama_index.core.postprocessor import LongContextReorder
from llama_index.postprocessor.colbert_rerank import ColbertRerank 

llm = Groq(temperature = 0.7, model = "Mixtral-8x7b-32768")

def ReciprocalRerankFusionRetriever(nodes, embedding, LLM, k = 2, nq = 4): 
  storage_context = StorageContext.from_defaults()
  storage_context.docstore.add_documents(nodes)

  index = VectorStoreIndex(nodes, storage_context = storage_context, embed_model = embedding)

  vector_retriever = index.as_retriever(similarity_top_k = k)

  bm25_retriever = BM25Retriever.from_defaults(
      docstore = index.docstore, similarity_top_k = k
  )
  
  retriever = QueryFusionRetriever(
    [vector_retriever, bm25_retriever],
    similarity_top_k = k,
    num_queries = nq,  # set this to 1 to disable query generation
    mode = "reciprocal_rerank",
    use_async = True,
    verbose = True, 
    llm = LLM
    # query_gen_prompt="...",  # we could override the query generation prompt here
  )
  return retriever

def reranker(model, k, chunks, LLM, embed_model = "local:BAAI/bge-base-en-v1.5"):
  if k == 1:
    return chunks
  else: 
    if model == 'LongContextReorder':
      postprocessor = LongContextReorder()
    elif model == 'LLMRerank':
      service_context = ServiceContext.from_defaults(llm = LLM, embed_model = embed_model)
      postprocessor = LLMRerank(top_n = k, service_context = service_context)
    elif model == 'JinaRerank':  
      postprocessor = JinaRerank(top_n = k, model = "jina-reranker-v1-base-en", api_key = "jina_eba96965c69b4fd894f57138b933d4195yirZCsUcmqh6-7DeqSi1Uu9DaHg")
    else:
      postprocessor = ColbertRerank(top_n = 5, model = "colbert-ir/colbertv2.0", tokenizer = "colbert-ir/colbertv2.0", keep_retrieval_score = True)

  return postprocessor.postprocess_nodes(chunks)
  
def embeddings(name_model):
  if name_model == 'text-embedding-3-large':
      model = OpenAIEmbedding(model = 'text-embedding-3-large')
  else: 
      model = HuggingFaceEmbedding(name_model)
  return model 

In [42]:
embedding = embeddings("BAAI/bge-small-en-v1.5")
retriever = ReciprocalRerankFusionRetriever(nodes, embedding, LLM = llm, k = 3)
retriever

c:\Users\ebt1023\AppData\Local\Programs\Python\Python312\Lib\site-packages\huggingface_hub\file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


In [18]:
import nest_asyncio

nest_asyncio.apply()

In [45]:
chunks = retriever.retrieve('How can I get the properties of a fulfilment?')
re = reranker('LongContextReorder', 3, chunks, LLM = llm)
re

Generated queries:
1. "What are the methods to retrieve properties of a fulfillment in programming?"
2. "How to access details of a fulfillment object in e-commerce?"
3. "What are the steps to obtain fulfillment properties through API?"


[NodeWithScore(node=TextNode(id_='e3be284d-a59a-43c7-b565-5e01448c0817', embedding=None, metadata={'filename': 'Create a fulfilment to authorize a service _ Healthanea Documentation.html', 'extension': '.html'}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=[], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='a2a51a01-c3df-4b94-aa3b-1897daba7671', node_type=<ObjectType.DOCUMENT: '4'>, metadata={'filename': 'Create a fulfilment to authorize a service _ Healthanea Documentation.html', 'extension': '.html'}, hash='176daef87353e632b6b7e46bc77b6d8cf20f5ce1392401cc3694166a7935db26'), <NodeRelationship.PREVIOUS: '2'>: RelatedNodeInfo(node_id='9a5ecca9-70bb-4606-982b-71328cef7699', node_type=<ObjectType.TEXT: '1'>, metadata={'filename': 'Create a fulfilment to authorize a service _ Healthanea Documentation.html', 'extension': '.html'}, hash='a3ef0352b22e78b9a02ba9f38850cdfa8973835a26639e963f21c4f4ab5a904d'), <NodeRelationship.NEXT: '3'>: RelatedNodeInfo(nod

In [49]:
for i in range(3):
    print(f"**Retrieval {i + 1}: {re[i].text[:30]}")
    print(f"The scores is {re[i].score}")
    #print(f"**Retrieval {i + 1} tokens:** {tokens(pr2[0].text)}")
    print(f"**Document Retrieval {i + 1}:** {re[0].metadata['filename'][:-5]}")
    print("---")

**Retrieval 1: These can include actions perf
The scores is 0.09730301427815971
**Document Retrieval 1:** Create a fulfilment to authorize a service _ Healthanea Documentation
---
**Retrieval 2: Healthanea returns a <code>202
The scores is 0.08224925083729949
**Document Retrieval 2:** Create a fulfilment to authorize a service _ Healthanea Documentation
---
**Retrieval 3: See <a class="xref" href="../a
The scores is 0.08306010928961748
**Document Retrieval 3:** Create a fulfilment to authorize a service _ Healthanea Documentation
---
